py/server/tests/test_keyed_record_adapter.py

```diff
@@ -0,0 +1,301 @@
+        self.assertEquals(1.1, aapl_trade.price)
+        self.assertEquals(10_000_000_000, aapl_trade.size)
+        self.assertEquals('ARCA', aapl_trade.exch)
+
+        aapl_trade = records[1]
+        self.assertEquals('AAPL', aapl_trade.sym)
+        self.assertEquals(1.2, aapl_trade.price)
+        self.assertEquals(1_000, aapl_trade.size)
+        self.assertEquals('NASDAQ', aapl_trade.exch)
+
+        self.assertEquals(None, records[2])
+
+    def testCustomKeyedRecordAdapterWithCompositeKeyColAndAdapter(self):
+        # source = Table(_JTstTools.testRefreshingTable(
+        #     _JTstTools.i(2, 4).copy().toTracking(),
+        #     _JTableTools.intCol("Sym", 'AAPL', 'AAPL', None),
+        #     _JTableTools.doubleCol("Price", 1.1, 1.2, NULL_DOUBLE),
+        #     _JTableTools.longCol("Size", 10_000_000_000, 1_000, NULL_LONG),
+        #     _JTableTools.col("Exch", "ARCA", 'NASDAQ', None),
+        # ))
+
+        source = new_table([
+            string_col("Sym", ['AAPL', 'AAPL', None]),
+            double_col("Price", [1.1, 1.2, NULL_DOUBLE]),
+            long_col("Size", [10_000_000_000, 1_000, NULL_LONG]),
+            string_col("Exch", ["ARCA", 'NASDAQ', None]),
+        ])
+        _JTableTools.show(source.j_table)
+
+        class StockTrade:
+            sym: str
+            price: float
+            size: int
+            exch: str
+
+            def __init__(self, sym: str, price: float, size: int, exch: str):
+                self.sym = sym
+                self.price = price
+                self.size = size
+                self.exch = exch
+
+        # Function that just calls StockTrades() with args in correct order (Exch last; when calling
+        # this method it is second arg because keys come first):
+        def create_stock_trade(sym: str, exch: str, price: float, size: int):
+            return StockTrade(sym, price, size, exch)
+
+        keyed_record_adapter: KeyedRecordAdapter[List[str], StockTrade] = make_record_adapter(
+            source,
+            create_stock_trade,
+            ['Sym', 'Exch'],
+            ["Price", "Size"]
+        )
+
+        records = keyed_record_adapter.get_records([
+            ["AAPL", "ARCA"],
+            ["AAPL", "NASDAQ"],
+            ["AAPL", "BADKEY"],
+        ])
+
+        self.assertEquals(list, type(records))
+        self.assertEquals(3, len(records))
+
+        aapl_trade = records[0]
+        self.assertEquals('AAPL', aapl_trade.sym)
+        self.assertEquals(1.1, aapl_trade.price)
+        self.assertEquals(10_000_000_000, aapl_trade.size)
+        self.assertEquals('ARCA', aapl_trade.exch)
+
+        aapl_trade = records[1]
+        self.assertEquals('AAPL', aapl_trade.sym)
+        self.assertEquals(1.2, aapl_trade.price)
+        self.assertEquals(1_000, aapl_trade.size)
+        self.assertEquals('NASDAQ', aapl_trade.exch)
+
+        self.assertEquals(None, records[2])
+
+
+if __name__ == '__main__':
+    unittest.main()
```

Comment:
In PartitionedTables, @rcaudy has resolved the 1- vs many-key values problem.  This whole PR should strive for the same for a user experience.